### PR TITLE
Import courses from metadata to publisher app.

### PR DESCRIPTION
**Pull request by awais786**
**Management command for migrating data from course-metadata to publisher.**

1. For each course in course-metadata only add the canonical course-run.
2. Add card image url in the publisher-course-run model to save the metadata course-run card image.
3. Add the course even if the authoring organization does't have default roles.

https://openedx.atlassian.net/browse/ECOM-7809

Management command take start_id and end_id. Both are pk of course-metadata table. e.g
`./manage.py import_metadata_courses --start_id=1 --end_id=1600`

### CODE REVIEW

File: course_discovery/apps/publisher/dataloader/__init__.py [hunk 1]
```diff
@@ -0,0 +1 @@
+__author__ = 'apple'
```
Comment (waheedahmed) [inline course_discovery/apps/publisher/dataloader/__init__.py]: Remove this.

File: course_discovery/apps/publisher/management/__init__.py [hunk 2]
```diff
@@ -0,0 +1 @@
+__author__ = 'apple'
```
Comment (waheedahmed) [inline course_discovery/apps/publisher/management/__init__.py]: Remove this.

File: course_discovery/apps/publisher/management/commands/__init__.py [hunk 3]
```diff
@@ -0,0 +1 @@
+__author__ = 'apple'
```
Comment (waheedahmed) [inline course_discovery/apps/publisher/management/commands/__init__.py]: Remove this.

File: course_discovery/apps/publisher/dataloader/create_courses.py [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+
+def organizations_requirements(organizations, meta_data_course):
+    """ Before adding course make sure organization exists and has OrganizationExtension
+    object also.
+    """
+    if organizations.count() > 1:
+        logger.warning(
+            'Course has more than 1 organization. Course uuid is [%s].', meta_data_course.uuid
+        )
+        return False
```
Comment (waheedahmed) [inline course_discovery/apps/publisher/dataloader/create_courses.py]: Return `None` instead `False`

File: course_discovery/apps/publisher/dataloader/create_courses.py [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+
+    available_organization = organizations.first()
+
+    if not available_organization:
+        logger.warning(
+            'Course has no organization. Course uuid is [%s].', meta_data_course.uuid
+        )
+        return False
```
Comment (waheedahmed) [inline course_discovery/apps/publisher/dataloader/create_courses.py]: Same here, return `None` instead `False`

File: course_discovery/apps/publisher/dataloader/create_courses.py [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+
+    if not OrganizationExtension.objects.filter(organization=available_organization).exists():
+        logger.warning(
+            'OrganizationExtension does not contain any record for organization [%s], course [%s].',
+            available_organization.uuid, meta_data_course.uuid
+        )
+        return False
```
Comment (waheedahmed) [inline course_discovery/apps/publisher/dataloader/create_courses.py]: Same here, return `None` instead `False`

File: course_discovery/apps/publisher/dataloader/create_courses.py [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+        )
+
+    logger.info('Import course with id [%s], number [%s].', publisher_course.id, publisher_course.number)
+
+    # create canonical course-run against the course.
+    create_course_runs(meta_data_course, publisher_course)
+
+
+def create_course_runs(meta_data_course, publisher_course):
+        # create or update all metadata course runs.
```
Comment (waheedahmed) [inline course_discovery/apps/publisher/dataloader/create_courses.py]: Here you are creating only one course-run against the course but from your comment it looks like a course can have multiple course runs, is that right?

File: course_discovery/apps/publisher/dataloader/create_courses.py [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+            create_seats(canonical_course_run, publisher_course_run)
+
+        else:
+            logger.warning(
+                'Canonical course-run not found for metadata course [%s]', meta_data_course.uuid
+            )
+
+
+def create_seats(metadata_course_run, publisher_course_run):
+    # create or update all metadata course runs.
```
Comment (waheedahmed) [inline course_discovery/apps/publisher/dataloader/create_courses.py]: *seats

File: course_discovery/apps/publisher/management/commands/import_metadata_courses.py [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+
+        )
+
+        parser.add_argument(
+            '--end_id',
+            action='store',
+            dest='end_id',
+            default=None,
+            required=True,
+            help='To this id courses will be downloaded.'
```
Comment (waheedahmed) [inline course_discovery/apps/publisher/management/commands/import_metadata_courses.py]: *Imported

File: course_discovery/apps/publisher/management/commands/import_metadata_courses.py [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+    help = 'Import courses into publisher app.'
+
+    def add_arguments(self, parser):
+        parser.add_argument(
+            '--start_id',
+            action='store',
+            dest='start_id',
+            default=None,
+            required=True,
+            help='The Primary key value starting id to download the courses.'
```
Comment (waheedahmed) [inline course_discovery/apps/publisher/management/commands/import_metadata_courses.py]: *import

File: course_discovery/apps/publisher/management/commands/tests/__init__.py [hunk 11]
```diff
@@ -0,0 +1 @@
+__author__ = 'apple'
```
Comment (waheedahmed) [inline course_discovery/apps/publisher/management/commands/tests/__init__.py]: Remove this.

File: course_discovery/apps/publisher/models.py [hunk 12]
```diff
@@ -261,6 +261,10 @@ class CourseRun(TimeStampedModel, ChangedByMixin):
     video_language = models.ForeignKey(LanguageTag, null=True, blank=True, related_name='video_language')
     preview_url = models.URLField(null=True, blank=True)
 
+    # temporary field to save the canonical course run image. In 2nd script this url field
+    # will be used to download the image and save into course model --> course image.
+    card_image_url = models.URLField(null=True, blank=True, verbose_name='canonical course run image')
```
Comment (waheedahmed) [inline course_discovery/apps/publisher/models.py:269]: Instead of downloading and saving the files, can't we use this field and Image field both?

**Review by waheedahmed — CHANGES_REQUESTED** [review]

File: course_discovery/apps/publisher/dataloader/create_courses.py (hunk 7 above)
Comment (awais786) [reply course_discovery/apps/publisher/dataloader/create_courses.py]: We just need to create most recent course-run. Canonical one

File: course_discovery/apps/publisher/models.py (hunk 12 above)
Comment (awais786) [reply course_discovery/apps/publisher/models.py:269]: That can be discussed. But first we need to import this field.

File: course_discovery/apps/publisher/dataloader/create_courses.py (hunk 7 above)
Comment (awais786) [reply course_discovery/apps/publisher/dataloader/create_courses.py]: Comment Updated.

**Review by waheedahmed — APPROVED** [review]

**Comment by awais786** [comment]
@tasawernawaz @waheedahmed Please review.

**Comment by awais786** [comment]
@tasawernawaz please review.

File: course_discovery/apps/publisher/dataloader/create_courses.py [hunk 13]
```diff
@@ -0,0 +1,206 @@
...
+    else:
+        logger.warning(
+            'No seats found for course-run [%s].', metadata_course_run.uuid
+        )
+
+
+def assign_course_user_roles(course, organization):
+    # Assign the course user roles against each organization users.
+
+    # add default organization roles into course-user-roles
```
Comment (tasawernawaz) [inline course_discovery/apps/publisher/dataloader/create_courses.py]: remove duplicate comments

**Review by tasawernawaz — APPROVED** [review]

**Pull request merged**